KAFKA-12756: Update ZooKeeper to v3.6.3

rondagostino opened the pull request:
Updates ZooKeeper to v3.6.3 and adds some additional exercising of ZooKeeper in the corresponding upgrade system tests.

The ZooKeeper v3.6.3 dependency on `io.dropwizard.metrics:metrics-core` cannot be avoided as per https://issues.apache.org/jira/browse/ZOOKEEPER-4324.  All integration tests based on `ZooKeeperTestHarness` fail without the dropwizard metrics jar being on the CLASSPATH; system tests fail as well.  This patch adds the library as a dependency.

This patch also adds 2.8.0 to the system test docker and Vagrant images (it was missing) and fixes `KafkaVersionTest.test_multi_version()` due to sending `kafka-topics --zookeeper` to the wrong node in a 2-broker cluster where one broker is v0.8.2 (which can use `kafka-topics --zookeeper`) and the other is the latest version (which no longer can due to https://github.com/apache/kafka/pull/10457).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rondagostino committed:
KAFKA-12756: Update ZooKeeper to v3.6.3

ijuma commented on docs/upgrade.html:
```diff
@@ -21,6 +21,13 @@
+        zookeeper.properties file does not remove this dependency.  ZooKeeper will fail to start with
+        <code>java.lang.NoClassDefFoundError: com/codahale/metrics/Reservoir</code> if you do not have the above library
+        on your CLASSPATH.
+    </li>
```
Isn't this handled automatically for users?

rondagostino replied:
> Isn't this handled automatically for users

Yeah, good point, no need to discuss it here.  Will remove.

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -61,6 +61,7 @@ versions += [
   bcpkix: "1.66",
   checkstyle: "8.36.2",
   commonsCli: "1.4",
+  dropwizardMetrics: "3.2.5",
```
Does this conflict with the yammer metrics library we use in Kafka?

rondagostino replied:
> Does this conflict with the yammer metrics library 

It doesn't, no.  The yammer metrics library is called `metrics-core-2.2.0.jar` and has all classes underneath the `com.yammer.metrics` package.  The dropwizard library is called `metrics-core-3.2.5.jar` and has all classes underneath the `com.codahale.metrics` package.  It isn't ideal that the jar names differ only by the version, but there is no conflict.

ijuma replied:
Why are we using this old version? The latest version is 4.2.2.

rondagostino replied:
https://github.com/apache/zookeeper/blob/release-3.6.3/pom.xml#L367

```
<dropwizard.version>3.2.5</dropwizard.version>
```

ijuma replied:
3.7 uses a newer version though, have we checked if we have to stick with this old version? It will be harder to update  in the 3.0 series after that's released.

rondagostino replied:
> have we checked if we have to stick with this old version

No.  How would we do that check?  I can think of 2 ways: 1) Feels too risky: just take the new version and call it good if the system tests pass; 2) Ask in the Zk mailing list maybe?  Any other options?

> It will be harder to update in the 3.0 series after that's released

It feels to me like the straightforward thing to do is use the version that ZK uses, and there's no need to update the dropwizard version unless we update the ZK version.  And if we are sticking with ZK 3.6.x through the EOL of ZooKeeper with Kafka (unknown, but potentially the case) then there's no reason to change it sans some kind of CVE.  Maybe I'm missing something, though?

ijuma replied:
The straightforward way would be to check the version bump in the 3.7 branch, did that require code changes?

CVEs are the common reason to have to update these days.

rondagostino replied:
https://github.com/apache/zookeeper/commit/13fe0d0ffb9fd2c379b9b430aaaf9ee75acfceba

It's not the only version update that was made, but no code changes.

rondagostino replied:
https://github.com/apache/kafka/pull/10982 contains this change.  Will attache system test results there when tests complete.

rondagostino committed:
No need to mention details of new library dependency in upgrade notes

rondagostino committed:
Fix system test failing due to use of --zookeeper option

rondagostino committed:
Add v2.8.0 to Vagrant system test image

rondagostino committed:
Make it a test dependency and add to CLASSPATH in kafka-run-class

rondagostino commented:
System test results before making the jar test dependency: https://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2021-06-23--001.1624469608--rondagostino--KAFKA-12756--bbc88d7e8/report.html
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.1
session_id:       2021-06-23--001
run time:         175 minutes 4.253 seconds
tests run:        867
passed:           659
failed:           11
ignored:          197
================================================================================
```

- `kafkatest.tests.core.zookeeper_authorizer_test` failed for `metadata_quorum:REMOTE_KRAFT` -- unrelated to this change snce ZooKeeper isn't even involved.
- `kafkatest.tests.core.upgrade_test` failed all 3 tests from version 2.8.0; this was due to 2.8.0 not being part of the Vagrant image.  I added a commit to this PR to fix that and confirmed that the non-compression flavor of that test for 2.8.0 passed locally with Vagrant (this PR already had the change to add it to the Docker image for Docker-based system tests).
- `kafkatest.sanity_checks.test_kafka_version` failed due to `kafka-topics --zookeeper` being used against the current version rather than the 0.8.2 version; added a commit to this PR to swap the node versions so the correct broker will get the request and confirmed locally that the test now passes.
- `kafkatest.tests.streams.streams_eos_test` had 4 test failures.
- `kafkatest.tests.streams.streams_upgrade_test` had 2 test failures

I believe the steams failures are unrelated because I downgraded to Zookeeper v3.5.9 locally and ran `kafkatest.tests.streams.streams_upgrade_test.StreamsUpgradeTest.test_version_probing_upgrade` -- it still failed.

rondagostino committed:
Revert to full runtime dependency

rondagostino commented:
I've reverted the change that made the `io.dropwizard.metrics:metrics-core` library a test-only dependency due to packaging concerns -- would it be included in the release, for example?

showuon commented on tests/docker/Dockerfile:
```diff
@@ -62,6 +62,7 @@ RUN mkdir -p "/opt/kafka-2.4.1" && chmod a+rw /opt/kafka-2.4.1 && curl -s "$KAFK
 RUN mkdir -p "/opt/kafka-2.5.1" && chmod a+rw /opt/kafka-2.5.1 && curl -s "$KAFKA_MIRROR/kafka_2.12-2.5.1.tgz" | tar xz --strip-components=1 -C "/opt/kafka-2.5.1"
 RUN mkdir -p "/opt/kafka-2.6.2" && chmod a+rw /opt/kafka-2.6.2 && curl -s "$KAFKA_MIRROR/kafka_2.12-2.6.2.tgz" | tar xz --strip-components=1 -C "/opt/kafka-2.6.2"
 RUN mkdir -p "/opt/kafka-2.7.1" && chmod a+rw /opt/kafka-2.7.1 && curl -s "$KAFKA_MIRROR/kafka_2.12-2.7.1.tgz" | tar xz --strip-components=1 -C "/opt/kafka-2.7.1"
+RUN mkdir -p "/opt/kafka-2.8.0" && chmod a+rw /opt/kafka-2.8.0 && curl -s "$KAFKA_MIRROR/kafka_2.12-2.8.0.tgz" | tar xz --strip-components=1 -C "/opt/kafka-2.8.0"
```
Thanks for adding this!

showuon reviewed (approved):
@rondagostino , thanks for fixing the system tests. LGTM!

I also confirmed that upgrade to zk v3.6.x doesn't need additional work.

> The upgrade from 3.5.7 to 3.6.0 can be executed as usual, no particular additional upgrade procedure is needed

REF: https://zookeeper.apache.org/releases.html

cmccabe commented on build.gradle:
```diff
@@ -823,6 +823,7 @@ project(':core') {
     implementation libs.scalaLogging
     implementation libs.slf4jApi
     implementation(libs.zookeeper) {
+      implementation libs.dropwizardMetrics
```
Can you add a comment explaining that this is needed for Apache ZooKeeper, and should NOT be used by Kafka itself?

rondagostino replied:
Good point -- added.

cmccabe reviewed (approved):
LGTM after the one comment I made is addressed.

Thanks, @rondagostino !

It's too bad that this adds a dependency, but as we discussed, there isn't really any way to avoid it without causing a lot of other problems.

rondagostino committed:
Add comment about dependency not to be used in Kafka code

cmccabe commented:
Thanks, @rondagostino . Since the only change was to a comment, I will not wait for Jenkins to re-run.
